packages/navigator/src/browser/navigator-widget.tsx

```diff
@@ -88,21 +93,55 @@ export class FileNavigatorWidget extends FileTreeWidget {
                         this.model.expandNode(child);
                     }
                 }
+            }),
+            this.workspaceService.onWorkspaceChanged(async event => {
+                await this.refreshWidget(event.roots, event.workspaceConfigFile);
             })
         ]);
     }
 
-    protected initialize(): void {
-        this.workspaceService.root.then(async resolvedRoot => {
-            if (resolvedRoot) {
-                const uri = new URI(resolvedRoot.uri);
+    protected async initialize(): Promise<void> {
+        const workspaceDescriptionData = await this.workspaceService.workspace;
+        await this.refreshWidget(workspaceDescriptionData.roots, workspaceDescriptionData.workspaceConfigFile);
+    }
+
+    protected async refreshWidget(roots: FileStat[], workspaceConfig: FileStat | undefined): Promise<void> {
```

Comment:
I think it should be renamed to `updateRoot` and moved to a navigator model, also simplified:
```ts
    protected updatedRoot(): void {
        this.model.root = this.createRoot();
    }
    protected createRoot(): Promise<TreeNode | undefined> {
        const workspace = await this.workspaceService.workspace;
        if (workspace.roots.length > 0) {
        ...
    }
```